### ISSUE TITLE
jax.random.wrap_key_data(impl=key.impl) problems

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#17758]
### Description

Problem: I would expect the following to work, in order to programmatically wrap/unwrap keys, but it does not

```python
b = jax.random.key(1)
jax.random.wrap_key_data(jax.random.key_data(b), impl=b.impl)
```

as it fails with error
```python
File ~/Documents/pythonenvs/netket/python-3.11.2/lib/python3.11/site-packages/jax/_src/random.py:139, in resolve_prng_impl(impl_spec)
    136   return PRNG_IMPLS[impl_spec]
    138 keys_fmt = ', '.join(f'"{s}"' for s in PRNG_IMPLS.keys())
--> 139 raise ValueError(f'unrecognized PRNG implementation "{impl_spec}". '
    140                  f'Did you mean one of: {keys_fmt}?')

ValueError: unrecognized PRNG implementation "fry". Did you mean one of: "threefry2x32", "rbg", "unsafe_rbg"?
```

### What jax/jaxlib version are you using?

jax==master (today, 25 September) jaxlib==0.4.16

### Which accelerator(s) are you using?

CPU

### Additional system info

MacOs

**Comment by jakevdp** [comment]
Hi, thanks for the report!

The `impl` argument to `wrap_random_key` must be a string, and `key.impl` is not a string. I suppose we could allow the `impl` argument to be a private `Impl` object, but this is an internal class that is not user-constructable, so that might add some confusion. What do you think would be best?

**Comment by jakevdp** [comment]
Thanks – I agree that that's a clean solution at runtime, but people love their types. I'm not sure how to annotate that meaningfully without either (1) making the type `Any`, (2) exposing the currently private `PRNGImpl` type which we'd prefer not be in the public API, or (3) keeping it as-is and requiring you to write `# type: ignore` when you write the code you suggest.  What do you think?

**Comment by PhilipVinc** [comment]
Well, I would like anything you define to be semi-public and semi-documented such that we don't have to change our code often. 
If you do not want to expose the `PRNGImpl` type, there might be an alternative.

I just noticed that `tag` is used to construct the `hash` of the `PRNGImpl` object, so it already must be 'correctly' defined.
Can't you instead change the (semi-public, `jax.extend`) API to register a new `PRNGImplementation` to register it under the `tag` name? Then, the fact that the implementations are stored in a dictionary is just an implementation detail.

In short, I'm suggesting adding a function 

```python
PRNG_IMPLS = {}
def register_prng_implementation(impl: PRNGImpl):
    if impl.tag not in PRNG_IMPLS:      
         PRNG_IMPLS[impl.tag] = impl
    else:
        error...
```

instead of exposing `PRNG_IMPLS` in `jax.extend.random` you could expose this function (as well as use it internally whenever you register your implementations).

**Comment by jakevdp** [comment]
Stepping back – we originally did not plan to expose `wrap_random_key`, because we expect usage of it to be relatively rare in the eventual steady state. Why do you need to unwrap and rewrap keys? Can you not change your package to work directly with the new-style keys?

**Comment by PhilipVinc** [comment]
To get `mpi` working, the bare minimum we need is to get the pointer to the data and the 'data size' . 
I could try to just treat an extended data type as a `uint` array of length `dtype.itemsize` as well...

**Comment by jakevdp** [comment]
Thanks for the feedback and discussion – I think #17768 should address this.

**Comment by froystig** [comment]
I commented on #17768 about the possibility of exposing a function, `jax.random.key_impl` that supports the round trip.

Separately, we should probably look into hiding the `impl` attribute for parity with other jax arrays.

> If you do not want to expose the `PRNGImpl` type, there might be an alternative.

We expose the `PRNGImpl` type via `jax.extend.random`, but not through the main jax API.

**Comment by froystig** [comment]
I believe #17983 (which replaced #17768) closes this. Please reopen if you disagree!
